src/Bridge/Doctrine/MongoDbOdm/Paginator.php

```diff
@@ -0,0 +1,163 @@
+use ApiPlatform\Core\Exception\InvalidArgumentException;
+use Doctrine\ODM\MongoDB\Iterator\Iterator;
+use Doctrine\ODM\MongoDB\UnitOfWork;
+
+/**
+ * Decorates the Doctrine MongoDB ODM paginator.
+ *
+ * @experimental
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+final class Paginator implements \IteratorAggregate, PaginatorInterface
+{
+    /**
+     * @var Iterator
+     */
+    private $mongoDbOdmIterator;
+    /**
+     * @var array
+     */
+    private $pipeline;
+    /**
+     * @var UnitOfWork
+     */
+    private $unitOfWork;
+    /**
+     * @var string
+     */
+    private $resourceClass;
+
+    /** @var \ArrayIterator */
+    private $iterator;
+
+    /**
+     * @var int
+     */
+    private $firstResult;
+    /**
+     * @var int
+     */
+    private $maxResults;
+    /**
+     * @var int
+     */
+    private $totalItems;
+
+    public function __construct(Iterator $mongoDbOdmIterator, UnitOfWork $unitOfWork, string $resourceClass, array $pipeline)
+    {
+        $this->mongoDbOdmIterator = $mongoDbOdmIterator;
+        $this->unitOfWork = $unitOfWork;
+        $this->resourceClass = $resourceClass;
+        $this->pipeline = $pipeline;
+
+        $resultsFacetInfo = $this->getFacetInfo('results');
+        $this->getFacetInfo('count');
+
+        // See https://github.com/alcaeus/mongo-php-adapter#mongocommandcursor
+        // Since the method getCursorInfo in CommandCursor always returns 0 for 'skip' and 'limit',
+        // the values set in the facet stage are used instead.
+        $this->firstResult = $this->getStageInfo($resultsFacetInfo, '$skip');
+        $this->maxResults = $this->getStageInfo($resultsFacetInfo, '$limit');
+        $this->totalItems = $mongoDbOdmIterator->toArray()[0]['count'][0]['count'] ?? 0;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCurrentPage(): float
+    {
+        return floor($this->firstResult / $this->maxResults) + 1.;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getLastPage(): float
+    {
+        return ceil($this->totalItems / $this->maxResults) ?: 1.;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItemsPerPage(): float
+    {
+        return (float) $this->maxResults;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTotalItems(): float
+    {
+        return (float) $this->totalItems;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getIterator(): \Traversable
+    {
+        return $this->iterator ?? $this->iterator = new \ArrayIterator(array_map(function ($result) {
+            return $this->unitOfWork->getOrCreateDocument($this->resourceClass, $result);
+        }, $this->mongoDbOdmIterator->toArray()[0]['results']));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function count(): int
+    {
+        return iterator_count($this->getIterator());
```

Comment:
This will load and hydrate all documents in order to run a `count`. You might want to avoid that and count raw data instead:
```
return count($this->mongoDbOdmIterator->toArray()[0]['results']);
```